core/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java

```diff
@@ -306,6 +311,13 @@ public void readFromWithId(long id, StreamInput in) throws IOException {
             serviceTimeEWMA = -1;
             nodeQueueSize = -1;
         }
+        if (in.getVersion().onOrAfter(Version.V_6_1_0)) {
+            String[] unmappedFieldsArray = in.readOptionalStringArray();            
+            unmappedFields = unmappedFieldsArray == null ? null : new HashSet<>(Arrays.asList(unmappedFieldsArray));
+        } else {
+            unmappedFields = null;
```

Comment:
could we not make this an empty set and avoid null checks?

Reply:
True but at the cost of adding some inefficiency.
We'd be allocating empty arrays for serialization of these empty sets then, when deserializing we're allocating HashSet objects to load these empty arrays (unless we add special-case code for spotting size 0 arrays and using empty set which is effectively as ugly as the existing null checks?)

Reply:
Ok fair enough, I'm happy leaving this as is then

Reply:
Actually I'd still prefer to go with Colin's idea to use empty sets. We can still optimize later by making sure to use a Collections.emptySet (which is a singleton) if the size is 0.